storage/provider/openstack.go

```diff
@@ -0,0 +1,328 @@
+func (s *openstackVolumeSource) DestroyVolumes(volIds []string) []error {
+
+	errors := make([]error, len(volIds))
+	for volIdx, volId := range volIds {
+		if err := s.openstackClient.DeleteVolume(volId); err != nil {
+			errors[volIdx] = err
+		}
+	}
+	return errors
+}
+
+// ValidateVolumeParams implements storage.VolumeSource.
+func (s *openstackVolumeSource) ValidateVolumeParams(params storage.VolumeParams) error {
+	panic("not implemented")
+}
+
+// AttachVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) AttachVolumes(args []storage.VolumeAttachmentParams) ([]storage.VolumeAttachment, error) {
+
+	volAttachments := make([]storage.VolumeAttachment, len(args))
+	for argIdx, attachArg := range args {
+		if attachArg.Provider != OpenstackProviderType {
+			logger.Debugf(
+				"openstack storage provider was passed a request for another provider type: %v",
+				attachArg.Provider,
+			)
+			continue
+		}
+
+		attachment, err := s.openstackClient.AttachVolume(
+			string(attachArg.InstanceId),
+			attachArg.VolumeId,
+			"", // TODO(katco): Fill this in when attachment support is added to parameters.
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		volAttachments[argIdx] = attachment
+	}
+
+	return volAttachments, nil
+}
+
+// DetachVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) DetachVolumes(params []storage.VolumeAttachmentParams) error {
+	panic("not implemented")
+}
+
+type OpenstackAdapter interface {
+	GetVolumesSimple() ([]storage.Volume, error)
+	DeleteVolume(string) error
+	CreateVolume(size uint64, name string) (storage.Volume, error)
+	AttachVolume(serverId, volumeId, mountPoint string) (storage.VolumeAttachment, error)
+}
+
+func newGooseAdapter(cfg *storage.Config) (OpenstackAdapter, error) {
+
+	region, _ := cfg.ValueString("region")
+
+	authClient := client.NewClient(&identity.Credentials{
+		URL:        cfg.Attrs()[openstackCfgKeyUrl].(string),
+		User:       cfg.Attrs()[openstackCfgKeyUser].(string),
+		Secrets:    cfg.Attrs()[openstackCfgKeySecrets].(string),
+		Region:     region,
+		TenantName: cfg.Attrs()[openstackCfgKeyTenantName].(string),
+	}, identity.AuthUserPass, nil)
+
+	endpoint := authClient.EndpointsForRegion(region)["volume"]
+	endpointUrl, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, errors.Annotate(err, "error parsing endpoint")
+	}
+	cinderClient := cinder.Basic(endpointUrl, authClient.TenantId(), authClient.Token)
+
+	novaClient := nova.New(authClient)
+
+	return &gooseAdapter{cinderClient, novaClient}, nil
+}
+
+type gooseAdapter struct {
+	cinder *cinder.Client
+	nova   *nova.Client
+}
+
+func (ga *gooseAdapter) CreateVolume(size uint64, name string) (storage.Volume, error) {
+
+	resp, err := ga.cinder.CreateVolume(cinder.CreateVolumeVolumeParams{
+		// NOTE: The Cinder documentation incorrectly states the
+		// size parameter is in GB. It is actually GiB.
+		Size: int(math.Ceil(float64(size / 1024))),
```

Comment:
This needs to be rounded up ie (size+1023)/1024

There are helper functions in providers/ec2 - mibToGib() and gibToMib()
Let's re-home those and use everywhere.